src/client/datascience/raw-kernel/rawJupyterSession.ts

```diff
@@ -200,35 +200,44 @@ export class RawJupyterSession extends BaseJupyterSession {
                 stacktrace
             });
             if (session) {
-                return (session as RawSession).kernelProcess.dispose();
+                return session.kernelProcess.dispose();
             }
         });
     }
 
-    protected setSession(session: RawSession) {
+    protected setSession(session: RawSession | undefined) {
         super.setSession(session);
-
-        // When setting the session clear our current exit handler and hook up to the
-        // new session process
-        if (this.processExitHandler) {
-            this.processExitHandler.dispose();
-            this.processExitHandler = undefined;
+        if (!session) {
+            return;
         }
-        if (session?.kernelProcess) {
-            // Watch to see if our process exits
-            this.processExitHandler = session.kernelProcess.exited(({ exitCode, reason }) => {
-                sendTelemetryEvent(Telemetry.RawKernelSessionKernelProcessExited, undefined, {
-                    exitCode,
-                    exitReason: getTelemetrySafeErrorMessageFromPythonTraceback(reason)
-                });
-                traceError(`Raw kernel process exited code: ${exitCode}`);
-                this.shutdown().catch((reason) => {
-                    traceError(`Error shutting down jupyter session: ${reason}`);
-                });
-                // Next code the user executes will show a session disposed message
+        this.isTerminating = undefined;
+        // Watch to see if our process exits
+        // This is the place to do this, after this session has been setup as the active kernel.
+        const disposable = session.kernelProcess.exited(({ exitCode, reason }) => {
+            // If this session is no longer the active session, then we don't need to do anything
+            // with this exit event (could be we're killing it, or restarting).
+            // In the case of restarting, the old session is disposed & a new one created.
+            // When disposing the old kernel we shouldn't fire events about session getting terminated.
+            if (session !== this.session) {
+                return;
+            }
+            sendTelemetryEvent(Telemetry.RawKernelSessionKernelProcessExited, undefined, {
+                exitCode,
+                exitReason: getTelemetrySafeErrorMessageFromPythonTraceback(reason)
             });
-            this._disposables.push(this.processExitHandler);
-        }
+            traceError(`Raw kernel process exited code: ${exitCode}`);
+            // If the raw kernel process dies, then send the terminating event, and shutdown the session.
+            // Afer shutting down the session, the status changes to `dead`
+            this.isTerminating = true;
+            this.onStatusChangedEvent.fire('terminating');
+            this.shutdown()
```

Comment:
Will fix this later in another PR, i dont think the session should be disposing iteself here.
But lets look into that later.